src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,893 @@
+  {
+    out << id << ", ";
+  }
+  out << "}\n";
+
+  out << "// top_level_tracked: {";
+  for(auto &id : top_level_tracked)
+  {
+    out << id << ", ";
+  }
+  out << "}\n";
+
+  out << "// loop:\n";
+  for(auto &loop : loop_info_map)
+  {
+    out << "// dfcc-loop_id:" << loop.first << "\n";
+    auto head = loop.second.find_head(goto_function.body);
+    auto latch = loop.second.find_latch(goto_function.body);
+    out << "// head:\n";
+    head.value()->output(out);
+    out << "// latch:\n";
+    latch.value()->output(out);
+    loop.second.output(out, ns, mode);
+  }
+  out << "// program:\n";
+  Forall_goto_program_instructions(target, goto_function.body)
+  {
+    out << "// dfcc-loop-id:" << dfcc_get_loop_id(target).value();
+    out << " cbmc-loop-number:" << target->loop_number;
+    out << " top-level:" << dfcc_is_loop_top_level(target);
+    out << " head:" << dfcc_is_loop_head(target);
+    out << " body:" << dfcc_is_loop_body(target);
+    out << " exiting:" << dfcc_is_loop_exiting(target);
+    out << " latch:" << dfcc_is_loop_latch(target);
+    out << "\n";
+    target->output(out);
+  }
+}
+
+const exprt &
+dfcc_cfg_infot::get_write_set(goto_programt::const_targett target) const
+{
+  auto loop_id_opt = dfcc_get_loop_id(target);
+  PRECONDITION(
+    loop_id_opt.has_value() &&
+    is_valid_loop_or_top_level_id(loop_id_opt.value()));
+  auto loop_id = loop_id_opt.value();
+  if(is_top_level_id(loop_id))
+  {
+    return top_level_write_set;
+  }
+  else
+  {
+    return loop_info_map.at(loop_id).addr_of_write_set_var;
+  }
+}
+
+const symbol_exprt &
+dfcc_cfg_infot::get_loop_write_set(goto_programt::const_targett target) const
+{
+  auto loop_id_opt = dfcc_get_loop_id(target);
+  PRECONDITION(
+    loop_id_opt.has_value() && is_valid_loop_id(loop_id_opt.value()));
+  auto loop_id = loop_id_opt.value();
+  return loop_info_map.at(loop_id).addr_of_write_set_var;
+}
+
+const exprt &dfcc_cfg_infot::get_outer_write_set(std::size_t loop_id) const
+{
+  PRECONDITION(is_valid_loop_id(loop_id));
+  auto outer_loop_opt = get_outer_loop_identifier(loop_id);
+  return outer_loop_opt.has_value()
+           ? get_loop_info(outer_loop_opt.value()).addr_of_write_set_var
+           : top_level_write_set;
+}
+
+const dfcc_loop_infot &
+dfcc_cfg_infot::get_loop_info(const std::size_t loop_id) const
+{
+  PRECONDITION(is_valid_loop_id(loop_id));
+  return loop_info_map.at(loop_id);
+}
+
+// find the identifier or the immediately enclosing loop in topological order
+const optionalt<std::size_t>
+dfcc_cfg_infot::get_outer_loop_identifier(const std::size_t loop_id) const
+{
+  PRECONDITION(is_valid_loop_id(loop_id));
+  auto outer_loops = get_loop_info(loop_id).outer_loops;
+
+  // find the first loop in the topological order that is connected
+  // to our node.
+  for(const auto &idx : get_loops_toposorted())
+  {
+    if(
+      std::find(outer_loops.begin(), outer_loops.end(), idx) !=
+      outer_loops.end())
+    {
+      return idx;
+    }
```

Comment:
It looks like this is the only read attempt of `outer_loops`. If so, wouldn't a `std::set` be a more appropriate data structure than a `std::vector` (nevermind the discussion whether it shouldn't be `std::optional` instead).

Reply:
Yes.